pkg/controller.v1/mpi/mpijob_controller.go

```diff
@@ -135,6 +135,11 @@ func (jc *MPIJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
+	if manager := jc.ManagedByExternalController(mpijob.Spec.RunPolicy); manager != nil {
```

Comment:
Instead of implementing this skip mechanism for every reconciler, could you implement a predictor?

I think we can implement predictors something like `OnJobCreateFunc` for all Jobs here: https://github.com/kubeflow/training-operator/blob/master/pkg/common/util/reconciler.go.

And then, we can specify the predictor here (this is a PyTorch example): https://github.com/kubeflow/training-operator/blob/c990a4ab846776a1f32e8241dee728aa012d23a5/pkg/controller.v1/pytorch/pytorchjob_controller.go#L182.

The Pod predicator is a good example of a predicator implemented in the TrainingOperator.
- https://github.com/kubeflow/training-operator/blob/e9766d171507ad3772cc04c740acb4d1ee4c544c/pkg/common/util/reconciler.go#L49
- https://github.com/kubeflow/training-operator/blob/c990a4ab846776a1f32e8241dee728aa012d23a5/pkg/controller.v1/pytorch/pytorchjob_controller.go#L200-L204

Reply:
There is only one mechanism, but called in each reconciler.
How this will differ if I have to setup the predicate in every controller?
What is the benefit of using predicate?
We want to reject the job even before the Reconcile?

Reply:
> We want to reject the job even before the Reconcile?

Yes, that my suggestion's motivation.
In a large cluster with hundreds or thousands of Jobs, reconciling performance is an important factor.
By using predicates, we can avoid the enqueue Jobs with managedBy to reconcile queue, and then the approach allows us to avoid performance regressions.

Reply:
I think I need different predicate for each KubeflowJob as the runPolicy is part of the Spec, the same as `onOwnerCreateFunc`.

> Instead of implementing this skip mechanism for every reconciler, could you implement a predictor?

@tenzen-y  This suggests me you thought differently. 
Wdyt now?

Reply:
We had this discussion when developing in the core k8s, we decided to go for a simple skip in reconciler.

One needs to be careful doing this only in preconditions. Consider the following:
1. user creates a `job-1` without `managedBy`, a sync is queued based on name / namespace
2. the `job-1` is deleted and shortly after `job-2` with the same name and namespace is created, but with `managedBy`
3. the queued reconcile executes and we reconcile `job-2` even though we should not

For this scenario we added this test which was reproducing the potential bug: https://github.com/kubernetes/kubernetes/blob/746f08a8da0a45e8b1a46362822db05cb11ed294/test/integration/job/job_test.go#L2209.

Still, we skip in reconciler + preconditions, but maybe this would be premature optimizing it.

Reply:
I had great discussions with @mimowo offline.
Throughout the discussions, we agreeded to implement this managedBy verification mechanism in the reconciler instead of the predicator.

If we implement this verification mechanism only in the predicator, there are some potential bugs in the situation where the Jobs are quickly replaced with the same namespacedName Job with managedBy field, as we discussed k/k: https://github.com/kubernetes/kubernetes/pull/123273#discussion_r1496378689

But, we can improve reconcile queue size by implementing this mechanism both in reconciler and predicator.
However, as a first step, we decided not to implement the predicator, and we will collect user feedback related to controller performance.